### PR TITLE
[Instrumentation.Hangfire] Added support for custom job display names

### LINKED ISSUE #755: [Instrumentation.Hangfire] Add support for custom job display names
# Issue with OpenTelemetry.Instrumentation.Hangfire

**Is this a feature request or a bug?**

* [X] Feature Request
* [ ] Bug

**What is the expected behavior?**

The Hangfire Dashboard provides a feature to display custom job display names. I would like to use the custom name for the display name of the job activity aswell.

**What is the actual behavior?**

The activity display name is always formatted as `{Job.Type.Name}.{Job.Method.Name}`

## Additional Context

In our project we use a single entry point for all jobs. To distinguish the different jobs in the Dashboard we display a custom job name using the `HangfireDashboardOptions.DisplayNameFunc`.

### PR DESCRIPTION
**Pull request by prochnowc**
Fixes #755.

## Changes

Adds support for using custom display names in the activities display name.

### CODE REVIEW

**Comment by Kielek** [comment]
@fred2u, could you please check?

**Comment by Kielek** [comment]
@prochnowc, for sure we need entry in https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.Hangfire/CHANGELOG.md

File: src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs [hunk 1]
```diff
@@ -56,7 +59,10 @@ public void OnPerforming(PerformingContext performingContext)
 
         if (activity != null)
         {
-            activity.DisplayName = $"JOB {performingContext.BackgroundJob.Job.Type.Name}.{performingContext.BackgroundJob.Job.Method.Name}";
+            Func<BackgroundJob, string> displayNameFunc =
+                this.options.DisplayNameFunc ?? HangfireInstrumentation.DefaultDisplayNameFunc;
+
+            activity.DisplayName = $"JOB {displayNameFunc(performingContext.BackgroundJob)}";
```
Comment (fred2u) [inline src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs:65]: Could you please move the prefix "JOB " in the DefaultDisplayNameFunc

**Comment by prochnowc** [comment]
> @prochnowc, for sure we need entry in https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.Hangfire/CHANGELOG.md

@Kielek I wasnt sure where to put it .. is an entry below "Unreleased" suitable?

**Review by fred2u — APPROVED** [review]

**Comment by Kielek** [comment]
> > @prochnowc, for sure we need entry in https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.Hangfire/CHANGELOG.md
> 
> @Kielek I wasnt sure where to put it .. is an entry below "Unreleased" suitable?

Yes, you can find example here: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/8f7fbd425c7750cc72c03633ff4bfd3d3aeedc42/src/OpenTelemetry.Contrib.Instrumentation.AWS/CHANGELOG.md?plain=1#L3-L8

**Comment by prochnowc** [comment]
> Yes, you can find example here:
> 
> https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/8f7fbd425c7750cc72c03633ff4bfd3d3aeedc42/src/OpenTelemetry.Contrib.Instrumentation.AWS/CHANGELOG.md?plain=1#L3-L8

@Kielek Ok, added the entry to CHANGELOG

**Review by fred2u — APPROVED** [review]

**Comment by prochnowc** [comment]
@fred2u The failing builds are related to `System.Diagnostics.DiagnosticSource 7.0.0-rc.1.22426.10` which complains about not supporting netcoreapp3.1 - I was already wondering why OpenTelemetry uses 7.0.0-rc instead of 6.0.0 ?

The build will probably fail for the same reason with v6.

**Comment by cijothomas** [comment]
> @fred2u The failing builds are related to `System.Diagnostics.DiagnosticSource 7.0.0-rc.1.22426.10` which complains about not supporting netcoreapp3.1 - I was already wondering why OpenTelemetry uses 7.0.0-rc instead of 6.0.0 ?
> 
> The build will probably fail for the same reason with v6.

^ That is only a warning and not an error. Build fails for different reason.

C:\Users\runneradmin\.nuget\packages\system.diagnostics.diagnosticsource\7.0.0-rc.1.22426.10\buildTransitive\netcoreapp2.0\System.Diagnostics.DiagnosticSource.targets(4,5): warning : System.Diagnostics.DiagnosticSource 7.0.0-rc.1.22426.10 doesn't support netcoreapp3.1 and has not been tested with it. Consider upgrading your TargetFramework to net6.0 or later. You may also set <SuppressTfmSupportBuildWarnings>true</SuppressTfmSupportBuildWarnings> in the project file to ignore this warning and attempt to run in this unsupported configuration at your own risk. [D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\test\OpenTelemetry.Instrumentation.Process.Tests\OpenTelemetry.Instrumentation.Process.Tests.csproj]
D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\src\OpenTelemetry.Instrumentation.Hangfire\HangfireInstrumentationOptions.cs(40,89): error SA1629: Documentation text should end with a period [D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\src\OpenTelemetry.Instrumentation.Hangfire\OpenTelemetry.Instrumentation.Hangfire.csproj]

**Comment by prochnowc** [comment]
> ^ That is only a warning and not an error. Build fails for different reason.

Oh, sorry - did not see the actual error because of the irritating warning. It's fixed now.

**Review by fred2u — APPROVED** [review]

**Review by Kielek — APPROVED** [review]

**Pull request merged**